radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/addressbook/PeerPostponingManager.java

```diff
@@ -0,0 +1,170 @@
+ * permissions under this License.
+ */
+
+package com.radixdlt.network.p2p.addressbook;
+
+import com.radixdlt.network.p2p.NodeId;
+import com.radixdlt.network.p2p.RadixNodeUri;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+
+/**
+ * Maintain information about peers connections to which should be postponed.
+ *
+ * <p>Class implements simple exponential backoff strategy, doubling postpone time for every
+ * recorded failure.
+ */
+final class PeerPostponingManager {
+  private final Map<RadixNodeUri, PostponeInfo> postponedPeers = new HashMap<>();
+  private final Clock clock;
+  private final Duration initialPostpone;
+  private final Duration maxPostpone;
+
+  public PeerPostponingManager(Clock clock, Duration initialPostpone, Duration maxPostpone) {
+    this.clock = clock;
+    this.initialPostpone = initialPostpone;
+    this.maxPostpone = maxPostpone;
+  }
+
+  public boolean recordFailure(RadixNodeUri uri) {
+    boolean survived = postponedPeers.compute(uri, this::computePostpone) != null;
+
+    if (!survived) {
+      removeAliases(uri.getNodeId());
+    }
+
+    return survived;
+  }
+
+  private void removeAliases(NodeId nodeId) {
+    postponedPeers.keySet().stream()
+        .filter(uri -> uri.getNodeId().equals(nodeId))
+        .toList() // "materialize" to avoid in place removal
+        .forEach(postponedPeers::remove);
+  }
+
+  private PostponeInfo computePostpone(RadixNodeUri ignoredKey, PostponeInfo postponeInfo) {
+    return postponeInfo == null ? new PostponeInfo() : checkSurvival(postponeInfo);
+  }
+
+  private static PostponeInfo checkSurvival(PostponeInfo postponeInfo) {
+    return postponeInfo.survivesFailure() ? postponeInfo : null;
+  }
+
+  public void recordSuccess(RadixNodeUri uri) {
+    postponedPeers.remove(uri);
+    // Remove other addresses which belong to same peer but already expired
+    cleanupExpired(uri);
+  }
+
+  private void cleanupExpired(RadixNodeUri uri) {
+    postponedPeers.entrySet().stream()
+        .filter(entry -> isExpiredAlias(uri.getNodeId(), entry))
+        .map(Entry::getKey)
+        .toList() // "materialize" to avoid in place removal
+        .forEach(postponedPeers::remove);
+  }
+
+  public boolean shouldIgnore(RadixNodeUri uri) {
+    return Optional.ofNullable(postponedPeers.get(uri))
+        .map(PostponeInfo::shouldIgnore)
+        .orElse(false);
+  }
+
+  private static boolean isExpiredAlias(NodeId nodeId, Entry<RadixNodeUri, PostponeInfo> entry) {
+    return entry.getKey().getNodeId().equals(nodeId) && entry.getValue().isExpired();
+  }
+
+  private final class PostponeInfo {
+    private final Instant firstFailure;
+    private Duration postponeDuration;
+
+    PostponeInfo() {
+      this.firstFailure = now();
+      this.postponeDuration = initialPostpone;
+    }
+
+    boolean survivesFailure() {
```

Comment:
This function has a side effect which was unexpected - I'd rather the side effect be clear from the method name (rather than the return statement) - or have separate methods: `void recordConnectionFailure()` and `bool shouldBeRemovedFromAddressBook()`

Reply:
This component needs to maintain some data which change over the time. 
Usually, there are two approaches - one you mention above, but it requires independent maintenance of the data, for example using dedicated thread (or similar concept, for example, time-based notifications). 
Another approach is to update time-based data when data are accessed. This method is the single point, which can be used for updates of internal state conveniently.

Reply:
I'm happy for a single method, but as per the comment above, I'd personally prefer the name to capture the side effect instead of the return value (`recordFailure` or `doublePostponeDuration`), or even more clearly/honestly, capture both: `doublePostponeDurationAndReturnIfShouldStayInAddressBook`